contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,1212 @@
+    error InvalidConsensusAddress();
+    // @notice signature: 0x3f259b7a
+    error UpdateTooFrequently();
+    // @notice signature: 0x5c32dd9c
+    error JailTimeNotExpired();
+    // @notice signature: 0xdc6f0bdd
+    error DelegationAmountTooSmall();
+    // @notice signature: 0x64689203
+    error OnlySelfDelegation();
+    // @notice signature: 0x9811e0c7
+    error ZeroShares();
+    // @notice signature: 0xf0e3e629
+    error SameValidator();
+    // @notice signature: 0xbd52fcdb
+    error NoMoreFelonyAllowed();
+    // @notice signature: 0x37233762
+    error AlreadySlashed();
+    // @notice signature: 0x90b8ec18
+    error TransferFailed();
+    // @notice signature: 0x41abc801
+    error InvalidRequest();
+    // @notice signature: 0x1898eb6b
+    error VoteAddressExpired();
+    // @notice signature: 0xc2aee074
+    error ConsensusAddressExpired();
+    // @notice signature: 0x0d7b78d4
+    error InvalidSynPackage();
+
+    /*----------------- storage -----------------*/
+    bool private _paused;
+    uint8 private _isReceivingFund;
+    uint256 public transferGasLimit;
+
+    // stake params
+    uint256 public minSelfDelegationBNB;
+    uint256 public minDelegationBNBChange;
+    uint256 public maxElectedValidators;
+    uint256 public unbondPeriod;
+    uint256 public redelegateFeeRate;
+
+    // slash params
+    uint256 public downtimeSlashAmount;
+    uint256 public felonySlashAmount;
+    uint256 public downtimeJailTime;
+    uint256 public felonyJailTime;
+
+    // validator operator address set
+    EnumerableSet.AddressSet private _validatorSet;
+    // validator operator address => validator info
+    mapping(address => Validator) private _validators;
+    // validator moniker set(hash of the moniker)
+    mapping(bytes32 => bool) private _monikerSet;
+    // validator consensus address => validator operator address
+    mapping(address => address) public consensusToOperator;
+    // validator consensus address => expiry date
+    mapping(address => uint256) public consensusExpiration;
+    // validator vote address => validator operator address
+    mapping(bytes => address) public voteToOperator;
+    // validator vote address => expiry date
+    mapping(bytes => uint256) public voteExpiration;
+
+    // legacy addresses of BC
+    mapping(address => bool) private _legacyConsensusAddress;
+    mapping(bytes => bool) private _legacyVoteAddress;
+
+    // total number of current jailed validators
+    uint256 public numOfJailed;
+    // max number of jailed validators between breathe block(only for malicious vote and double sign)
+    uint256 public maxFelonyBetweenBreatheBlock;
+    // index(timestamp / breatheBlockInterval) => number of malicious vote and double sign slash
+    mapping(uint256 => uint256) private _felonyMap;
+    // slash key => slash jail time
+    mapping(bytes32 => uint256) private _felonyRecords;
+
+    address public assetProtector;
+    mapping(address => bool) public blackList;
```

Comment:
Can we share a black list? Stake, Gov, TokenHub all have black list. For a same bad address, we need to add several times.

Reply:
Done